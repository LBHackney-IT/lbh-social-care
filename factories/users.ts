import { Factory } from 'fishery';

import { User } from 'types';

export const userFactory = Factory.define<User>(() => ({
  name: 'foo',
  hasAdminPermissions: true,
  hasChildrenPermissions: true,
  hasAdultPermissions: true,
  email: 'foo@bar.com',
  permissionFlag: 'A',
  isAuthorised: true,
}));

export const mockedUser = userFactory.build();
