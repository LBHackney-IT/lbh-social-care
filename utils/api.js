import { getQueryString } from 'utils/urls';

export const getInfiniteKey = (baseUrl, key, params) => (
  pageIndex,
  previousPageData
) => {
  if (previousPageData && !previousPageData[key]) return null;
  if (pageIndex === 0)
    return `${baseUrl}${params ? `?${getQueryString(params)}` : ''}`;
  return `${baseUrl}?${getQueryString({
    ...params,
    cursor: previousPageData.nextCursor,
  })}`;
};
