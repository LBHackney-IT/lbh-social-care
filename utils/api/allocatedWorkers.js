import useSWR from 'swr';
import axios from 'axios';

export const getAllocatedWorkers = (id) =>
  useSWR(`/api/residents/${id}/allocations`);

export const getResidentAllocation = (id, allocationId) =>
  useSWR(`/api/residents/${id}/allocations/${allocationId}`);

export const getTeams = ({ ageContext } = {}) =>
  useSWR(`/api/teams${ageContext ? '?ageContext=' + ageContext : ''}`);

export const getTeamWorkers = (teamId) =>
  useSWR(teamId ? `/api/teams/${teamId}/workers` : null);

export const getAllocationsByWorker = (workerId) =>
  useSWR(`/api/workers/${workerId}/allocations`);

export const deleteAllocatedWorker = async (residentId, body) => {
  const { data } = await axios.patch(
    `/api/residents/${residentId}/allocations`,
    body
  );
  return data;
};

export const addAllocatedWorker = async (residentId, body) => {
  const { data } = await axios.post(
    `/api/residents/${residentId}/allocations`,
    body
  );
  return data;
};
