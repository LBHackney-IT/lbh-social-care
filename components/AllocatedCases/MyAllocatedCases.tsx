import AllocatedCasesTable from 'components/AllocatedCases/AllocatedCasesTable';
import ErrorMessage from 'components/ErrorMessage/ErrorMessage';
import { useMyData } from 'utils/api/me';
import Spinner from 'components/Spinner/Spinner';

const MyAllocatedCases: React.FC = () => {
  const { data, error } = useMyData();
  if (error) {
    if (error?.response?.status === 404) {
      return (
        <p className="govuk-body govuk-!-margin-top-5">
          No people are assigned to you
        </p>
      );
    }
    return <ErrorMessage />;
  }
  if (!data) {
    return <Spinner />;
  }
  return (
    <>
      {data.allocations && (
        <>
          {data.allocations?.length > 0 ? (
            <>
              <p>
                Displaying ({data.allocations.length}){' '}
                {data.allocations.length > 1 ? 'allocations' : 'allocation'}
              </p>
              <AllocatedCasesTable cases={data.allocations} />
            </>
          ) : (
            <p className="govuk-body govuk-!-margin-top-5">
              No people are assigned to you
            </p>
          )}
        </>
      )}
    </>
  );
};

export default MyAllocatedCases;
