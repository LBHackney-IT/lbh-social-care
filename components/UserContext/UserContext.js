import { createContext, useContext, useEffect, useState } from 'react';
import { useRouter } from 'next/router';

import { AUTH_WHITELIST } from 'utils/auth';
import { getUser } from 'utils/api/checkAuth';
import Layout from 'components/Layout';
import Spinner from 'components/Spinner/Spinner';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState();
  const [redirect, setRedirect] = useState();
  const [isLoading, setLoading] = useState(true);
  const router = useRouter();
  useEffect(() => {
    const initializeAuth = async () => {
      try {
        const data = getUser();
        setUser(data);
      } catch (e) {
        setUser();
        setRedirect(e.response.status === 403 ? '/access-denied' : '/login');
      }
      setLoading(false);
    };
    initializeAuth();
  }, []);
  if (isLoading || !router.isReady) {
    return (
      <Layout>
        <Spinner />
      </Layout>
    );
  }
  if (
    typeof window !== 'undefined' &&
    AUTH_WHITELIST.includes(router.pathname) === Boolean(user)
  ) {
    router.push(redirect || location);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
        isLoading,
        setUser,
        redirect,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
