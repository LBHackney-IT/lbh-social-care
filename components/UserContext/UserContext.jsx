import { createContext, useContext } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children, user }) => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
