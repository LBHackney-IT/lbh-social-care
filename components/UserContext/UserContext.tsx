import { createContext, useContext, ReactNode, ReactElement } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

import type { User } from 'types';

interface Props {
  children: ReactNode;
  user?: User;
}

export const UserContext = createContext<{ user?: User }>({
  user: undefined,
});

export const AuthProvider = ({ children, user }: Props): ReactElement => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export const useAuth = (): { user?: User } => {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
};
