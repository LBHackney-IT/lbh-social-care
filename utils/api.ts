import { getQueryString } from 'utils/urls';

interface PageData {
  nextCursor?: string;
  [key: string]: unknown;
}

export const getInfiniteKey = (
  baseUrl: string,
  key: string,
  params: Record<string, unknown>
) => (pageIndex: number, previousPageData: PageData): string | null => {
  if (previousPageData && !previousPageData[key]) return null;
  if (pageIndex === 0)
    return `${baseUrl}${params ? `?${getQueryString(params)}` : ''}`;
  return `${baseUrl}?${getQueryString({
    ...params,
    cursor: previousPageData.nextCursor,
  })}`;
};
