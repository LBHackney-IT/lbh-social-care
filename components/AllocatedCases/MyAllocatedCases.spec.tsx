import { render } from '@testing-library/react';

import { UserContext } from 'components/UserContext/UserContext';
import { userFactory } from 'factories/users';
import * as allocatedWorkersAPI from 'utils/api/allocatedWorkers';
import { mockedAllocations } from 'factories/allocatedWorkers';
import MyAllocatedCases from './MyAllocatedCases';

jest.mock('components/Spinner/Spinner', () => () => 'MockedSpinner');

jest.mock('components/AllocatedCases/AllocatedCasesTable', () => () =>
  'MockedAllocatedCasesTable'
);

describe(`MyAllocatedCases`, () => {
  jest
    .spyOn(allocatedWorkersAPI, 'useMyAllocations')
    .mockImplementation(() => ({
      data: {
        allocations: mockedAllocations,
      },
      revalidate: jest.fn(),
      mutate: jest.fn(),
      isValidating: false,
    }));

  it('should render properly', async () => {
    const { findByText, asFragment } = render(
      <UserContext.Provider
        value={{
          user: userFactory.build(),
        }}
      >
        <MyAllocatedCases />
      </UserContext.Provider>
    );
    const allocateTable = await findByText('MockedAllocatedCasesTable');
    expect(allocateTable).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
