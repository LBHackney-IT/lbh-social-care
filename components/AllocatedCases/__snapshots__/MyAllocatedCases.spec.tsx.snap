// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyAllocatedCases should render an error if the API returns a 500 1`] = `
<DocumentFragment>
  <span
    class="govuk-error-message"
  >
    <span
      class="govuk-visually-hidden"
    >
      Error:
    </span>
     Oops an error occurred
  </span>
</DocumentFragment>
`;

exports[`MyAllocatedCases should render nothing when there is no worker for the current user 1`] = `<DocumentFragment />`;

exports[`MyAllocatedCases should render properly 1`] = `
<DocumentFragment>
  <p>
    Displaying (1) allocation
  </p>
  MockedAllocatedCasesTable
</DocumentFragment>
`;
